x-pack/plugins/security_solution/server/endpoint/services/actions/action_details_by_id.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core/server';
+import { getActionCompletionInfo, mapToNormalizedActionRequest } from './utils';
+import {
+  ActionDetails,
+  ActivityLogAction,
+  ActivityLogActionResponse,
+  EndpointAction,
+  EndpointActionResponse,
+  EndpointActivityLogAction,
+  EndpointActivityLogActionResponse,
+  LogsEndpointAction,
+  LogsEndpointActionResponse,
+} from '../../../../common/endpoint/types';
+import {
+  ACTION_REQUEST_INDICES,
+  ACTION_RESPONSE_INDICES,
+  catchAndWrapError,
+  categorizeActionResults,
+  categorizeResponseResults,
+  getUniqueLogData,
+} from '../../utils';
+import { EndpointError } from '../../../../common/endpoint/errors';
+import { NotFoundError } from '../../errors';
+import { ACTIONS_SEARCH_PAGE_SIZE } from './constants';
+
+export const getActionDetailsById = async (
+  esClient: ElasticsearchClient,
+  actionId: string
+): Promise<ActionDetails> => {
+  let actionRequestsLogEntries: Array<ActivityLogAction | EndpointActivityLogAction>;
+
+  let normalizedActionRequest: ReturnType<typeof mapToNormalizedActionRequest> | undefined;
+  let actionResponses: Array<ActivityLogActionResponse | EndpointActivityLogActionResponse>;
+
+  try {
+    // Get both the Action Request(s) and action Response(s)
+    const [actionRequestEsSearchResults, actionResponsesEsSearchResults] = await Promise.all([
+      // Get the action request(s)
+      esClient
+        .search<EndpointAction | LogsEndpointAction>(
+          {
+            index: ACTION_REQUEST_INDICES,
+            body: {
+              query: {
+                bool: {
+                  filter: [{ term: { action_id: actionId } }],
```

Comment:
need to add 1-2 more filters here. Whichever field has `'endpoint'` as its value. I've forgotten which key it is, something like `input_type` or weirdly named like that. But for us the value is always the constant string `'endpoint'` because it's telling agent the destination process.

We need that because the action_ids are not required to be UUIDs, they are opaque any strings. And they are not unique across plugins/clients.

Reply:
I assume the Endpoint index has `input_type` as alias field? (I'll try it out in the dev tools)